source/common/secret/sds_api.cc

```diff
@@ -154,14 +168,17 @@ void SdsApi::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reaso
   init_target_.ready();
 }
 
-absl::Status SdsApi::validateUpdateSize(int num_resources) {
-  if (num_resources == 0) {
+absl::Status SdsApi::validateUpdateSize(uint32_t added_resources_num,
+                                        uint32_t removed_resources_num) {
+  if (added_resources_num + removed_resources_num == 0) {
     return absl::InvalidArgumentError(
         fmt::format("Missing SDS resources for {} in onConfigUpdate()", sds_config_name_));
   }
-  if (num_resources != 1) {
+  if (added_resources_num > 1 || removed_resources_num > 1) {
     return absl::InvalidArgumentError(
-        fmt::format("Unexpected SDS secrets length: {}", num_resources));
+        fmt::format("Unexpected SDS secrets length for {}, number of added resources "
+                    "{}, number of removed resources {}",
+                    sds_config_name_, added_resources_num, removed_resources_num));
   }
   return absl::OkStatus();
```

Comment:
hmm... this seems to allow adding and removing of the resource (which I guess should just be updating the resource) when both added and removed are 1.
To the best of my (limited) understanding SDS should currently be a single-resource subscription. That implies that for a given resource name it should either be in the added part or removed part.

If you think this is incorrect, please point me to a test or config that you are aware of.

Reply:
That was my understanding as well based on the code, but I came across this [test](https://github.com/envoyproxy/envoy/blob/ffcc257e16c9046b2fec7497a6bf9293d8ada286/test/common/secret/sds_api_test.cc#L562) that seemed to imply add + removal is allowed? I'm open to modifying the test though (with a release note clarifying the behavior)

Reply:
Interesting. I don't have enough context on this test.
@htuch seems the test was added [here](https://github.com/envoyproxy/envoy/commit/73c46c9aba2c032f8fccb32e1b2c0c7a1df25a8d#diff-8e5f56463d7c4e0a70b111fef53b4e4d535b3c690b901441834fc4c891989167R138-R166)  (that you've reviewed and maybe you have some context), and we are wondering whether an xDS response of a single resource in delta-xDS can have both add/update and remove included?

Reply:
I don't think this was really deliberate AFAIU. It's been a long time so I don't have much context to add, but it seems a case where you could go either way with behaviors (ignore an unrelated or nonsensical removal, or perhaps a strong error like a NACK), this seems reasonable.

Reply:
Yeah testing that we don't NACK makes sense. I'm more inclined to keep this behavior for stability's sake and since the spec tends to be permissive of nonsensical removals

Reply:
I think this behavior is fine to retain, but please add a comment explaining what's going on for posterity.